File: Progs/libMax/maxbot-js/src/router.ts
import { Context } from './context';
import type { Client } from './client';
import type { Update } from './types';

export type Handler = (ctx: Context) => Promise<void> | void;
export type Middleware = (next: Handler) => Handler;

export class Router {
  private readonly commands = new Map<string, Handler>();
  private onText?: Handler;
  private onCallback?: Handler;
  private onAny?: Handler;
  private readonly middlewares: Middleware[] = [];

  use(mw: Middleware): void {
    this.middlewares.push(mw);
  }

  handleCommand(command: string, handler: Handler): void {
    const cmd = normalizeCommand(command);
    if (!cmd) return;
    this.commands.set(cmd, handler);
  }

  handleText(handler: Handler): void {
    this.onText = handler;
  }

  handleCallback(handler: Handler): void {
    this.onCallback = handler;
  }

  handleAny(handler: Handler): void {
    this.onAny = handler;
  }

  async dispatch(client: Client, update: Update): Promise<void> {
    const ctx = new Context(client, update);

    if (update.message) {
      const cmd = ctx.command();
      if (cmd && this.commands.has(cmd)) {
        await runChain(this.middlewares, this.commands.get(cmd) ?? noop, ctx);
        return;
      }
      if (this.onText) {
        await runChain(this.middlewares, this.onText, ctx);
        return;
      }
      if (this.onAny) {
        await runChain(this.middlewares, this.onAny, ctx);
      }
      return;
    }

    if (update.callback_query) {
      if (this.onCallback) {
        await runChain(this.middlewares, this.onCallback, ctx);
        return;
      }
      if (this.onAny) {
        await runChain(this.middlewares, this.onAny, ctx);
      }
    }
  }
}

function runChain(middlewares: Middleware[], handler: Handler, ctx: Context): Promise<void> {
  let current = handler;
  for (let i = middlewares.length - 1; i >= 0; i -= 1) {
    current = middlewares[i](current);
  }
  return Promise.resolve(current(ctx));
}

function normalizeCommand(command: string): string {
  return command.trim().replace(/^\//, '').toLowerCase();
}

const noop: Handler = () => undefined;

File: Progs/libMax/maxbot-go/router.go
package maxbot

import (
	"context"
	"strings"
)

type Handler func(*Context) error
type Middleware func(Handler) Handler

type Router struct {
	commands    map[string]Handler
	onText      Handler
	onCallback  Handler
	middlewares []Middleware
}

func NewRouter() *Router {
	return &Router{
		commands: make(map[string]Handler),
	}
}

func (r *Router) Use(mw Middleware) {
	r.middlewares = append(r.middlewares, mw)
}

func (r *Router) HandleCommand(cmd string, handler Handler) {
	cmd = strings.TrimPrefix(strings.TrimSpace(strings.ToLower(cmd)), "/")
	if cmd == "" || handler == nil {
		return
	}
	r.commands[cmd] = handler
}

func (r *Router) HandleText(handler Handler) {
	r.onText = handler
}

func (r *Router) HandleCallback(handler Handler) {
	r.onCallback = handler
}

func (r *Router) Dispatch(ctx context.Context, client *Client, upd Update) error {
	c := &Context{
		ctx:    ctx,
		Client: client,
		Update: upd,
	}
	if upd.Message != nil {
		if cmd := upd.Message.Command(); cmd != "" {
			if h, ok := r.commands[cmd]; ok {
				return chain(r.middlewares, h)(c)
			}
		}
		if r.onText != nil {
			return chain(r.middlewares, r.onText)(c)
		}
		return nil
	}
	if upd.Callback != nil && r.onCallback != nil {
		return chain(r.middlewares, r.onCallback)(c)
	}
	return nil
}

func chain(middlewares []Middleware, final Handler) Handler {
	if final == nil {
		return func(*Context) error { return nil }
	}
	h := final
	for i := len(middlewares) - 1; i >= 0; i-- {
		h = middlewares[i](h)
	}
	return h
}
